Fixed HD-1549 validate decompression path

altmannmarcelo opened the pull request:
Problem:
While decompressing a backup, we pass filepath and dest_filepath to
system call without any validation.

Fix:
Quote those path's in order to make it more secure.

altmannmarcelo committed:
Fixed HD-1549 validate decompression path

Problem:
While decompressing a backup, we pass filepath and dest_filepath to
system call without any validation.

Fix:
Quote those path's in order to make it more secure.

rahulmalik87 reviewed (approved):
LGTM

satya-bodapati commented on storage/innobase/include/xb0xb.h:
```diff
@@ -65,4 +65,5 @@ xb_fetch_tablespace_key(ulint space_id, byte *key, byte *iv);
 }
 #endif
 
+#define SQUOTE(str) "'" << str << "'"
```
Looks familiar :D

altmannmarcelo replied:
that is called backporting :D

Oneiroi reviewed (changes requested):
Please refer to the associated ticket that spawned this PR, the fix is noted as incomplete at this stage and needs further refinement to meet the intended purpose.

altmannmarcelo committed:
Addressed review comments

Enhanced the fix to also escape single quotes to block users getting out
of cmd.

Oneiroi commented on storage/innobase/xtrabackup/src/backup_copy.cc:
```diff
@@ -2219,10 +2219,16 @@ bool
 {
 	std::stringstream cmd, message;
-	char *dest_filepath = strdup(filepath);
+	char buf[FN_LEN];
```
Where is `FN_LEN` being defined ? 

I note this is present in https://github.com/percona/percona-xtrabackup/blob/da04f2363f6ed1f601a690cc56856131c6d574b8/include/my_io.h#L79 however this is not being included in this file, I've missed or am unable to identify where FN_LEN is being set (this is set as `256` in the my_io.h header file fwiw)

altmannmarcelo replied:
backup_copy.cc -> #include xtrabackup.h -> #include xbstream.h -> #include <my_io.h>

Oneiroi commented on storage/innobase/xtrabackup/src/backup_copy.cc:
```diff
@@ -2219,10 +2219,16 @@ bool
 	bool needs_action = false;
 
-	cmd << "cat " << SQUOTE(filepath);
+	if (escape_string_for_mysql(&my_charset_utf8_general_ci, buf, 0,
```
I presume this is the function that's being called here: https://github.com/percona/percona-xtrabackup/blob/d4cbe0e6d4b4895dda372353f3abab23aa17c96a/mysys/charset.cc#L733-L833

This is pretty exhaustive and even escapes multibyte chars afaict (read: this _may_ also intercept the use of unicode homoglyph's in filenames also). 👍

altmannmarcelo replied:
correct, also in that sense we might better use `my_charset_utf8mb4_general_ci` instead so we are covered on the 4th byte too.

Oneiroi reviewed (approved):
Please see my notes on `FN_LEN`, right now this LGTM though I am also aware that my COVID brain may be causing me to miss something so to err on the side of caution, I will request validation from stakeholders and communicate their comments via the HD ticket, thanks :+1:
